File: api/app/services/permission_service.py
"""
Permission service for role-based access control.
"""
from typing import List, Optional
from sqlalchemy.orm import Session

from app.services.base import BaseService
from app.repositories.permission_repository import (
    PermissionRepository,
    OrganizationRolePermissionRepository,
    SystemRoleConfigRepository,
    SystemRolePermissionRepository,
    UserSystemRoleRepository,
)
from app.repositories.organization_repository import OrganizationMemberRepository
from app.models.permission import (
    Permission,
    Resource,
    Action,
    SystemRole,
    OrganizationRolePermission,
    SystemRoleConfig,
    SystemRolePermission,
)
from app.models.organization import MemberRole
from app.utils.permission_defaults import (
    get_permission_name,
    DEFAULT_ORG_PERMISSIONS,
    DEFAULT_SYSTEM_ROLE_PERMISSIONS,
    SYSTEM_ROLE_CONFIGS,
)


class PermissionService(BaseService):
    """权限服务"""

    def __init__(self, db: Session):
        super().__init__(db)
        self.permission_repo = PermissionRepository(db)
        self.org_role_perm_repo = OrganizationRolePermissionRepository(db)
        self.system_role_config_repo = SystemRoleConfigRepository(db)
        self.system_role_perm_repo = SystemRolePermissionRepository(db)
        self.user_system_role_repo = UserSystemRoleRepository(db)
        self.member_repo = OrganizationMemberRepository(db)

    def initialize_permissions(self) -> None:
        """初始化权限数据（系统启动时调用）"""
        for resource in Resource:
            for action in Action:
                code = f"{resource.value}:{action.value}"
                if not self.permission_repo.find_by_code(code):
                    permission = Permission(
                        resource=resource,
                        action=action,
                        code=code,
                        name=get_permission_name(resource, action),
                    )
                    self.permission_repo.create(permission)

    def initialize_system_roles(self) -> None:
        """初始化系统角色配置"""
        for config in SYSTEM_ROLE_CONFIGS:
            if not self.system_role_config_repo.find_by_role(config["role"]):
                role_config = SystemRoleConfig(
                    role=config["role"],
                    name=config["name"],
                    description=config.get("description"),
                    is_active=True,
                )
                self.system_role_config_repo.create(role_config)

        # 初始化系统角色权限
        for role, perms in DEFAULT_SYSTEM_ROLE_PERMISSIONS.items():
            for resource, action in perms:
                permission = self.permission_repo.find_by_resource_action(
                    resource, action
                )
                if permission:
                    existing = (
                        self.db.query(SystemRolePermission)
                        .filter(
                            SystemRolePermission.role == role,
                            SystemRolePermission.permission_id == permission.id,
                        )
                        .first()
                    )
                    if not existing:
                        role_perm = SystemRolePermission(
                            role=role,
                            permission_id=permission.id,
                            is_enabled=True,
                        )
                        self.db.add(role_perm)
        self.db.commit()

    def initialize_org_permissions(self, org_id: int) -> None:
        """初始化组织默认权限配置"""
        for role, perms in DEFAULT_ORG_PERMISSIONS.items():
            for resource, action in perms:
                permission = self.permission_repo.find_by_resource_action(
                    resource, action
                )
                if permission:
                    # 检查是否已存在
                    existing = (
                        self.db.query(OrganizationRolePermission)
                        .filter(
                            OrganizationRolePermission.organization_id == org_id,
                            OrganizationRolePermission.role == role.value,
                            OrganizationRolePermission.permission_id == permission.id,
                        )
                        .first()
                    )
                    if not existing:
                        config = OrganizationRolePermission(
                            organization_id=org_id,
                            role=role.value,
                            permission_id=permission.id,
                            is_enabled=True,
                        )
                        self.db.add(config)
        self.db.commit()

    def check_permission(
        self, user_id: int, org_id: int, permission_code: str
    ) -> bool:
        """
        检查用户在组织中是否拥有指定权限

        优先级：
        1. 系统超级管理员拥有所有权限
        2. owner 角色拥有所有权限
        3. 检查角色权限配置
        """
        # 检查系统角色
        if self.user_system_role_repo.is_super_admin(user_id):
            return True

        # 获取用户在组织中的角色
        membership = self.member_repo.find_membership(org_id, user_id)
        if not membership:
            return False

        # owner 拥有所有权限
        if membership.role == MemberRole.OWNER:
            return True

        # 检查角色权限配置
        return self.org_role_perm_repo.has_permission(
            org_id, membership.role.value, permission_code
        )

    def check_system_permission(
        self, user_id: int, permission_code: str
    ) -> bool:
        """
        检查用户是否拥有系统级权限

        用于跨组织的权限检查
        """
        # 超级管理员拥有所有权限
        if self.user_system_role_repo.is_super_admin(user_id):
            return True

        # 获取用户的所有系统角色
        user_roles = self.user_system_role_repo.get_user_system_roles(user_id)

        # 检查任意一个系统角色是否有该权限
        for role in user_roles:
            if self.system_role_perm_repo.has_permission(role, permission_code):
                return True

        return False

    def get_role_permissions(
        self, org_id: int, role: MemberRole
    ) -> List[Permission]:
        """获取组织角色的所有启用的权限"""
        if role == MemberRole.OWNER:
            # owner 返回所有权限
            return self.permission_repo.get_all_permissions()

        return self.org_role_perm_repo.get_enabled_permissions(org_id, role.value)

    def update_role_permissions(
        self,
        org_id: int,
        role: MemberRole,
        permission_codes: List[str],
        user_id: int,
    ) -> bool:
        """
        更新组织角色权限

        owner 角色权限不可修改
        """
        # 只有 owner 可以修改权限
        if not self.member_repo.has_role(org_id, user_id, [MemberRole.OWNER]):
            return False

        # owner 权限不可修改
        if role == MemberRole.OWNER:
            return False

        # 获取权限ID
        permissions = self.permission_repo.get_permissions_by_codes(permission_codes)
        permission_ids = [p.id for p in permissions]

        self.org_role_perm_repo.set_role_permissions(org_id, role.value, permission_ids)
        return True

    def get_all_permissions(self) -> List[Permission]:
        """获取所有可用权限（用于UI展示）"""
        return self.permission_repo.get_all_permissions()

    def get_permissions_grouped(self) -> dict[str, List[Permission]]:
        """获取按资源分组的权限"""
        permissions = self.get_all_permissions()
        grouped: dict[str, List[Permission]] = {}
        for perm in permissions:
            if perm.resource.value not in grouped:
                grouped[perm.resource.value] = []
            grouped[perm.resource.value].append(perm)
        return grouped

    def get_user_permissions(self, user_id: int, org_id: int) -> List[str]:
        """获取用户在组织中的所有权限代码"""
        # 系统超级管理员返回所有权限
        if self.user_system_role_repo.is_super_admin(user_id):
            return [p.code for p in self.get_all_permissions()]

        membership = self.member_repo.find_membership(org_id, user_id)
        if not membership:
            return []

        # owner 返回所有权限
        if membership.role == MemberRole.OWNER:
            return [p.code for p in self.get_all_permissions()]

        # 返回角色启用的权限
        permissions = self.get_role_permissions(org_id, membership.role)
        return [p.code for p in permissions]

    def get_user_system_roles(self, user_id: int) -> List[SystemRole]:
        """获取用户的所有系统角色"""
        return self.user_system_role_repo.get_user_system_roles(user_id)

    def is_super_admin(self, user_id: int) -> bool:
        """检查用户是否是超级管理员"""
        return self.user_system_role_repo.is_super_admin(user_id)

    def get_system_role_configs(self) -> List[SystemRoleConfig]:
        """获取所有系统角色配置"""
        return self.system_role_config_repo.get_all_active()

    def grant_system_role(
        self, user_id: int, role: SystemRole, granted_by: int
    ) -> bool:
        """授予用户系统角色"""
        # 只有超级管理员可以授予系统角色
        if not self.is_super_admin(granted_by):
            return False

        self.user_system_role_repo.grant_role(user_id, role, granted_by)
        return True

    def revoke_system_role(
        self, user_id: int, role: SystemRole, revoked_by: int
    ) -> bool:
        """撤销用户的系统角色"""
        # 只有超级管理员可以撤销系统角色
        if not self.is_super_admin(revoked_by):
            return False

        return self.user_system_role_repo.revoke_role(user_id, role)

File: api/src/services/permission.service.ts
import type { OrgRole } from '@prisma/client';
import type { PermissionRepository } from '../repositories/permission.repo.js';
import { defaultPermissionRepo } from '../repositories/permission.repo.js';
import { createAppError } from '../utils/appError.js';
import { ulid } from 'ulid';
import {
  RESOURCES,
  ACTIONS,
  RESOURCE_NAMES,
  ACTION_NAMES,
} from '@apartment-ultra/api-contract';
import {
  type Resource,
  type Action,
} from '../constants/permissionDefaults.js';
import type { Prisma } from '@prisma/client';

/**
 * 权限响应
 */
export interface PermissionInfo {
  resource: Resource;
  action: Action;
  code: string;
  name: string;
}

/**
 * 分组权限响应
 */
export interface GroupedPermissions {
  [resource: string]: PermissionInfo[];
}

/**
 * 角色权限响应
 */
export interface RolePermissionsResult {
  id: string;
  name: string;
  description: string | null;
  is_system: boolean;
  permissions: string[];
}

/**
 * 角色列表项（带成员数）
 */
export interface OrgRoleListItem {
  id: string;
  name: string;
  description: string | null;
  is_system: boolean;
  permissions: string[];
  member_count: number;
}

/**
 * Permission Service 接口
 */
export interface PermissionService {
  // 权限列表（静态定义）
  listAll(): PermissionInfo[];
  listGrouped(): GroupedPermissions;

  // 组织角色
  listOrgRoles(orgId: string): Promise<OrgRoleListItem[]>;
  getOrgRole(roleId: string): Promise<OrgRole | null>;
  createOrgRole(orgId: string, name: string, description?: string): Promise<OrgRole>;
  updateOrgRole(
    roleId: string,
    data: { name?: string; description?: string; permissions?: string[] }
  ): Promise<OrgRole>;
  deleteOrgRole(roleId: string, requesterId: string): Promise<void>;
  getRolePermissions(roleId: string): Promise<RolePermissionsResult>;
  updateRolePermissions(
    roleId: string,
    permissionCodes: string[],
    requesterId: string
  ): Promise<void>;

  // 我的权限
  getMyPermissions(userId: string, orgId: string): Promise<string[]>;
}

/**
 * 创建 Permission Service 实例
 */
export function createPermissionService(
  getRepo: () => PermissionRepository = () => defaultPermissionRepo
): PermissionService {
  // 构建静态权限列表
  const allPermissions: PermissionInfo[] = [];
  for (const resource of RESOURCES) {
    for (const action of ACTIONS) {
      allPermissions.push({
        resource,
        action,
        code: `${resource}:${action}`,
        name: `${RESOURCE_NAMES[resource]}${ACTION_NAMES[action]}`,
      });
    }
  }

  return {
    /**
     * 获取所有权限列表（静态定义）
     */
    listAll: () => {
      return allPermissions;
    },

    /**
     * 获取分组权限列表
     */
    listGrouped: () => {
      const grouped: GroupedPermissions = {};
      for (const p of allPermissions) {
        if (!grouped[p.resource]) {
          grouped[p.resource] = [];
        }
        grouped[p.resource].push(p);
      }
      return grouped;
    },

    /**
     * 获取组织角色列表
     */
    listOrgRoles: async (orgId: string) => {
      const roles = await getRepo().findOrgRoles(orgId);
      const items: OrgRoleListItem[] = [];
      for (const role of roles) {
        const memberCount = await getRepo().countRoleMembers(role.id);
        items.push({
          id: role.id,
          name: role.name,
          description: role.description,
          is_system: role.is_system,
          permissions: (role.permissions as string[]) ?? [],
          member_count: memberCount,
        });
      }
      return items;
    },

    /**
     * 获取角色详情
     */
    getOrgRole: async (roleId: string) => {
      return getRepo().findOrgRoleById(roleId);
    },

    /**
     * 创建角色
     */
    createOrgRole: async (orgId: string, name: string, description?: string) => {
      // 检查是否已存在同名角色
      const existing = await getRepo().findOrgRoleByName(orgId, name);
      if (existing) {
        throw createAppError(400, '该角色名称已存在');
      }

      return getRepo().createOrgRole({
        id: ulid().toLowerCase(),
        organization_id: orgId,
        name,
        description,
        is_system: false, // 用户创建的角色不是系统预制
        permissions: [],
      });
    },

    /**
     * 更新角色
     */
    updateOrgRole: async (
      roleId: string,
      data: { name?: string; description?: string; permissions?: string[] }
    ) => {
      const role = await getRepo().findOrgRoleById(roleId);
      if (!role) {
        throw createAppError(404, '角色不存在');
      }

      if (role.is_system) {
        throw createAppError(403, '预制角色不可修改');
      }

      // 如果要更新名称，检查是否与现有角色冲突
      if (data.name && data.name !== role.name) {
        const existing = await getRepo().findOrgRoleByName(role.organization_id, data.name);
        if (existing) {
          throw createAppError(400, '该角色名称已存在');
        }
      }

      return getRepo().updateOrgRole(roleId, {
        name: data.name,
        description: data.description,
        permissions: data.permissions,
      } as { name?: string; description?: string; permissions?: Prisma.InputJsonValue });
    },

    /**
     * 删除角色
     */
    deleteOrgRole: async (roleId: string, _requesterId: string) => {
      const role = await getRepo().findOrgRoleById(roleId);
      if (!role) {
        throw createAppError(404, '角色不存在');
      }

      if (role.is_system) {
        throw createAppError(403, '预制角色不可删除');
      }

      // 检查是否有关联成员
      const memberCount = await getRepo().countRoleMembers(roleId);
      if (memberCount > 0) {
        throw createAppError(400, `该角色已关联 ${memberCount} 个账号，无法删除`);
      }

      await getRepo().deleteOrgRole(roleId);
    },

    /**
     * 获取角色权限
     */
    getRolePermissions: async (roleId: string) => {
      const role = await getRepo().findOrgRoleById(roleId);
      if (!role) {
        throw createAppError(404, '角色不存在');
      }

      return {
        id: role.id,
        name: role.name,
        description: role.description,
        is_system: role.is_system,
        permissions: (role.permissions as string[]) ?? [],
      };
    },

    /**
     * 更新角色权限
     */
    updateRolePermissions: async (
      roleId: string,
      permissionCodes: string[],
      _requesterId: string
    ) => {
      const role = await getRepo().findOrgRoleById(roleId);
      if (!role) {
        throw createAppError(404, '角色不存在');
      }

      if (role.is_system) {
        throw createAppError(403, '预制角色不可修改权限');
      }

      // 验证权限码是否有效
      const validCodes = new Set(allPermissions.map((p) => p.code));
      const codes = permissionCodes.filter((c) => validCodes.has(c));

      await getRepo().updateOrgRole(roleId, {
        permissions: codes as Prisma.InputJsonValue,
      });
    },

    /**
     * 获取我的权限
     */
    getMyPermissions: async (userId: string, orgId: string) => {
      const member = await getRepo().findMemberWithRole(orgId, userId);
      if (!member) {
        return [];
      }

      return (member.role.permissions as string[]) ?? [];
    },
  };
}

/**
 * 默认实例
 */
export const defaultPermissionService = createPermissionService();
